library/psa_crypto.c

```diff
@@ -3887,7 +3874,707 @@ psa_status_t psa_aead_decrypt( mbedtls_svc_key_id_t key,
+        || !operation->ad_started || !operation->body_started)
+    {
+        return( PSA_ERROR_BAD_STATE );
+    }
+
+    if( operation->lengths_set )
+    {
+        if( operation->ad_remaining != 0 || operation->body_remaining != 0 )
+        {
+            return( PSA_ERROR_BAD_STATE );
+        }
+    }
+
+    output_tag_length = PSA_AEAD_TAG_LENGTH(operation->alg);
+
+    if( tag_size < output_tag_length)
+    {
+        return ( PSA_ERROR_BUFFER_TOO_SMALL );
+    }
+
+    finish_output_size = PSA_AEAD_FINISH_OUTPUT_SIZE(operation->alg);
+
+    if( ciphertext_size < finish_output_size )
+    {
+        return ( PSA_ERROR_BUFFER_TOO_SMALL );
+    }
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_GCM)
+    if( operation->alg == PSA_ALG_GCM )
+    {
+        if( operation->buffer_length )
+        {
+
+            status =  mbedtls_to_psa_error( mbedtls_gcm_update( &operation->ctx.gcm,
+                                                                PSA_BLOCK_CIPHER_BLOCK_MAX_SIZE,
+                                                                operation->body_buffer,
+                                                                ciphertext ) );
+
+            if( status != PSA_SUCCESS )
+            {
+                return( status );
+            }
+
+            *ciphertext_length = finish_output_size;
+        }
+
+        status =  mbedtls_to_psa_error( mbedtls_gcm_finish( &operation->ctx.gcm,
+                                                            tag,
+                                                            tag_size ) );
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_GCM */
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)
+    if( operation->alg == PSA_ALG_CCM )
+    {
+        /* Multipart CCM not supported as yet */
+        ( void ) ciphertext;
+        ( void ) ciphertext_size;
+        ( void ) ciphertext_length;
+        ( void ) tag;
+        ( void ) tag_size;
+        ( void ) tag_length;
+
+        return ( PSA_ERROR_NOT_SUPPORTED );
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_CCM */
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_CHACHA20_POLY1305)
+    if( operation->alg == PSA_ALG_CHACHA20_POLY1305 )
+    {
+        status = mbedtls_to_psa_error( mbedtls_chachapoly_finish( &operation->ctx.chachapoly,
+                                                                  tag ) );
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_CHACHA20_POLY1305 */
+    {
+        ( void ) ciphertext;
+        ( void ) ciphertext_size;
+        ( void ) ciphertext_length;
+        ( void ) tag;
+        ( void ) tag_size;
+        ( void ) tag_length;
+
+        return ( PSA_ERROR_NOT_SUPPORTED );
+    }
+
+    if( status == PSA_SUCCESS )
+    {
+        *tag_length = output_tag_length;
+    }
+
+    psa_aead_abort_internal(operation);
+
+    return ( status );
+}
+
+/* Finish authenticating and decrypting a message in a multipart AEAD
+ * operation.*/
+psa_status_t psa_aead_verify(psa_aead_operation_t *operation,
```

Comment:
There's quite a lot of duplication between `finish` and `verify`. Please avoid code repetition and put the common parts in an auxiliary function.